File: claude_setup/domain/stack_mapping.py
from __future__ import annotations

from typing import Dict, List, Tuple

# --- Language + Build Tool -> Commands Mapping ---
# Key: (language, build_tool) tuple
# Value: dict with compile_cmd, build_cmd, test_cmd, coverage_cmd,
#        file_extension, build_file, package_manager

LANGUAGE_COMMANDS: Dict[Tuple[str, str], Dict[str, str]] = {
    ("java", "maven"): {
        "compile_cmd": "./mvnw compile -q",
        "build_cmd": "./mvnw package -DskipTests",
        "test_cmd": "./mvnw verify",
        "coverage_cmd": "./mvnw verify jacoco:report",
        "file_extension": ".java",
        "build_file": "pom.xml",
        "package_manager": "maven",
    },
    ("java", "gradle"): {
        "compile_cmd": "./gradlew compileJava -q",
        "build_cmd": "./gradlew build -x test",
        "test_cmd": "./gradlew test",
        "coverage_cmd": "./gradlew test jacocoTestReport",
        "file_extension": ".java",
        "build_file": "build.gradle",
        "package_manager": "gradle",
    },
    ("kotlin", "gradle"): {
        "compile_cmd": "./gradlew compileKotlin -q",
        "build_cmd": "./gradlew build -x test",
        "test_cmd": "./gradlew test",
        "coverage_cmd": "./gradlew test jacocoTestReport",
        "file_extension": ".kt",
        "build_file": "build.gradle.kts",
        "package_manager": "gradle",
    },
    ("typescript", "npm"): {
        "compile_cmd": "npx --no-install tsc --noEmit",
        "build_cmd": "npm run build",
        "test_cmd": "npm test",
        "coverage_cmd": "npm test -- --coverage",
        "file_extension": ".ts",
        "build_file": "package.json",
        "package_manager": "npm",
    },
    ("python", "pip"): {
        "compile_cmd": "python3 -m py_compile",
        "build_cmd": "pip install -e .",
        "test_cmd": "pytest",
        "coverage_cmd": "pytest --cov",
        "file_extension": ".py",
        "build_file": "pyproject.toml",
        "package_manager": "pip",
    },
    ("go", "go"): {
        "compile_cmd": "go build ./...",
        "build_cmd": "go build ./...",
        "test_cmd": "go test ./...",
        "coverage_cmd": "go test -coverprofile=coverage.out ./...",
        "file_extension": ".go",
        "build_file": "go.mod",
        "package_manager": "go",
    },
    ("rust", "cargo"): {
        "compile_cmd": "cargo check",
        "build_cmd": "cargo build",
        "test_cmd": "cargo test",
        "coverage_cmd": "cargo tarpaulin",
        "file_extension": ".rs",
        "build_file": "Cargo.toml",
        "package_manager": "cargo",
    },
    ("csharp", "dotnet"): {
        "compile_cmd": "dotnet build --no-restore --verbosity quiet",
        "build_cmd": "dotnet build",
        "test_cmd": "dotnet test",
        "coverage_cmd": 'dotnet test --collect:"XPlat Code Coverage"',
        "file_extension": ".cs",
        "build_file": "*.csproj",
        "package_manager": "dotnet",
    },
}

# --- Framework -> Default Port ---

FRAMEWORK_PORTS: Dict[str, int] = {
    "quarkus": 8080,
    "spring-boot": 8080,
    "nestjs": 3000,
    "express": 3000,
    "fastapi": 8000,
    "django": 8000,
    "gin": 8080,
    "ktor": 8080,
    "axum": 3000,
    "actix-web": 8080,
    "aspnet": 5000,
}

DEFAULT_PORT_FALLBACK = 8080

# --- Framework -> Health Path ---

FRAMEWORK_HEALTH_PATHS: Dict[str, str] = {
    "quarkus": "/q/health",
    "spring-boot": "/actuator/health",
    "nestjs": "/health",
    "express": "/health",
    "fastapi": "/health",
    "django": "/health",
    "gin": "/health",
    "ktor": "/health",
    "axum": "/health",
    "actix-web": "/health",
    "aspnet": "/health",
}

DEFAULT_HEALTH_PATH = "/health"

# --- Framework -> Valid Languages ---

FRAMEWORK_LANGUAGE_RULES: Dict[str, List[str]] = {
    "quarkus": ["java", "kotlin"],
    "spring-boot": ["java", "kotlin"],
    "nestjs": ["typescript"],
    "express": ["typescript"],
    "fastify": ["typescript"],
    "fastapi": ["python"],
    "django": ["python"],
    "flask": ["python"],
    "stdlib": ["go"],
    "gin": ["go"],
    "fiber": ["go"],
    "ktor": ["kotlin"],
    "axum": ["rust"],
    "actix-web": ["rust"],
    "aspnet": ["csharp"],
}

# --- Native Build Support ---

NATIVE_SUPPORTED_FRAMEWORKS: Tuple[str, ...] = (
    "quarkus",
    "spring-boot",
)

# --- Interface Types ---

VALID_INTERFACE_TYPES: Tuple[str, ...] = (
    "rest",
    "grpc",
    "graphql",
    "websocket",
    "tcp-custom",
    "cli",
    "event-consumer",
    "event-producer",
    "scheduled",
)

# --- Architecture Styles ---

VALID_ARCHITECTURE_STYLES: Tuple[str, ...] = (
    "microservice",
    "modular-monolith",
    "monolith",
    "library",
    "serverless",
)

# --- Interface Type -> Protocol ---

INTERFACE_PROTOCOL_MAP: Dict[str, str] = {
    "rest": "openapi",
    "grpc": "proto3",
    "graphql": "graphql",
    "websocket": "websocket",
    "tcp-custom": "tcp-custom",
    "event-consumer": "kafka",
    "event-producer": "kafka",
}

# --- Docker Base Images ---
# Template strings with {version} placeholder

DOCKER_BASE_IMAGES: Dict[str, str] = {
    "java": "eclipse-temurin:{version}-jre-alpine",
    "kotlin": "eclipse-temurin:{version}-jre-alpine",
    "typescript": "node:{version}-alpine",
    "python": "python:{version}-slim",
    "go": "golang:{version}-alpine",
    "rust": "rust:{version}-slim",
    "csharp": "mcr.microsoft.com/dotnet/aspnet:{version}",
}

DEFAULT_DOCKER_IMAGE = "alpine:latest"

File: src/domain/stack-mapping.ts
/**
 * Stack mapping constants and helpers.
 *
 * Migrated from Python `domain/stack_mapping.py`.
 * Key: `"language-build_tool"` string (Python used tuple keys).
 */

/** Command set for a language + build tool combination. */
export interface LanguageCommandSet {
  readonly compileCmd: string;
  readonly buildCmd: string;
  readonly testCmd: string;
  readonly coverageCmd: string;
  readonly fileExtension: string;
  readonly buildFile: string;
  readonly packageManager: string;
}

/** Language + build tool to command mapping (8 entries). */
export const LANGUAGE_COMMANDS: Readonly<Record<string, LanguageCommandSet>> = {
  "java-maven": {
    compileCmd: "./mvnw compile -q",
    buildCmd: "./mvnw package -DskipTests",
    testCmd: "./mvnw verify",
    coverageCmd: "./mvnw verify jacoco:report",
    fileExtension: ".java",
    buildFile: "pom.xml",
    packageManager: "maven",
  },
  "java-gradle": {
    compileCmd: "./gradlew compileJava -q",
    buildCmd: "./gradlew build -x test",
    testCmd: "./gradlew test",
    coverageCmd: "./gradlew test jacocoTestReport",
    fileExtension: ".java",
    buildFile: "build.gradle",
    packageManager: "gradle",
  },
  "kotlin-gradle": {
    compileCmd: "./gradlew compileKotlin -q",
    buildCmd: "./gradlew build -x test",
    testCmd: "./gradlew test",
    coverageCmd: "./gradlew test jacocoTestReport",
    fileExtension: ".kt",
    buildFile: "build.gradle.kts",
    packageManager: "gradle",
  },
  "typescript-npm": {
    compileCmd: "npx --no-install tsc --noEmit",
    buildCmd: "npm run build",
    testCmd: "npm test",
    coverageCmd: "npm test -- --coverage",
    fileExtension: ".ts",
    buildFile: "package.json",
    packageManager: "npm",
  },
  "python-pip": {
    compileCmd: "python3 -m py_compile",
    buildCmd: "pip install -e .",
    testCmd: "pytest",
    coverageCmd: "pytest --cov",
    fileExtension: ".py",
    buildFile: "pyproject.toml",
    packageManager: "pip",
  },
  "go-go": {
    compileCmd: "go build ./...",
    buildCmd: "go build ./...",
    testCmd: "go test ./...",
    coverageCmd: "go test -coverprofile=coverage.out ./...",
    fileExtension: ".go",
    buildFile: "go.mod",
    packageManager: "go",
  },
  "rust-cargo": {
    compileCmd: "cargo check",
    buildCmd: "cargo build",
    testCmd: "cargo test",
    coverageCmd: "cargo tarpaulin",
    fileExtension: ".rs",
    buildFile: "Cargo.toml",
    packageManager: "cargo",
  },
  "csharp-dotnet": {
    compileCmd: "dotnet build --no-restore --verbosity quiet",
    buildCmd: "dotnet build",
    testCmd: "dotnet test",
    coverageCmd: 'dotnet test --collect:"XPlat Code Coverage"',
    fileExtension: ".cs",
    buildFile: "*.csproj",
    packageManager: "dotnet",
  },
};

/** Framework to default port (11 entries). */
export const FRAMEWORK_PORTS: Readonly<Record<string, number>> = {
  "quarkus": 8080,
  "spring-boot": 8080,
  "nestjs": 3000,
  "express": 3000,
  "fastapi": 8000,
  "django": 8000,
  "gin": 8080,
  "ktor": 8080,
  "axum": 3000,
  "actix-web": 8080,
  "aspnet": 5000,
};

export const DEFAULT_PORT_FALLBACK = 8080;

/** Framework to health check path (11 entries). */
export const FRAMEWORK_HEALTH_PATHS: Readonly<Record<string, string>> = {
  "quarkus": "/q/health",
  "spring-boot": "/actuator/health",
  "nestjs": "/health",
  "express": "/health",
  "fastapi": "/health",
  "django": "/health",
  "gin": "/health",
  "ktor": "/health",
  "axum": "/health",
  "actix-web": "/health",
  "aspnet": "/health",
};

export const DEFAULT_HEALTH_PATH = "/health";

/** Framework to valid languages (15 entries). */
export const FRAMEWORK_LANGUAGE_RULES: Readonly<Record<string, readonly string[]>> = {
  "quarkus": ["java", "kotlin"],
  "spring-boot": ["java", "kotlin"],
  "nestjs": ["typescript"],
  "express": ["typescript"],
  "fastify": ["typescript"],
  "fastapi": ["python"],
  "django": ["python"],
  "flask": ["python"],
  "stdlib": ["go"],
  "gin": ["go"],
  "fiber": ["go"],
  "ktor": ["kotlin"],
  "axum": ["rust"],
  "actix-web": ["rust"],
  "aspnet": ["csharp"],
};

/** Frameworks that support native/GraalVM builds. */
export const NATIVE_SUPPORTED_FRAMEWORKS: readonly string[] = [
  "quarkus",
  "spring-boot",
];

/** Valid interface types (9 entries). */
export const VALID_INTERFACE_TYPES: readonly string[] = [
  "rest",
  "grpc",
  "graphql",
  "websocket",
  "tcp-custom",
  "cli",
  "event-consumer",
  "event-producer",
  "scheduled",
];

/** Valid architecture styles (5 entries). */
export const VALID_ARCHITECTURE_STYLES: readonly string[] = [
  "microservice",
  "modular-monolith",
  "monolith",
  "library",
  "serverless",
];

/** Interface type to protocol spec name (e.g. openapi, proto3). */
export const INTERFACE_SPEC_PROTOCOL_MAP: Readonly<Record<string, string>> = {
  "rest": "openapi",
  "grpc": "proto3",
  "graphql": "graphql",
  "websocket": "websocket",
  "tcp-custom": "tcp-custom",
  "event-consumer": "kafka",
  "event-producer": "kafka",
};

/** Docker base image templates with `{version}` placeholder. */
export const DOCKER_BASE_IMAGES: Readonly<Record<string, string>> = {
  "java": "eclipse-temurin:{version}-jre-alpine",
  "kotlin": "eclipse-temurin:{version}-jre-alpine",
  "typescript": "node:{version}-alpine",
  "python": "python:{version}-slim",
  "go": "golang:{version}-alpine",
  "rust": "rust:{version}-slim",
  "csharp": "mcr.microsoft.com/dotnet/aspnet:{version}",
};

export const DEFAULT_DOCKER_IMAGE = "alpine:latest";

/** Hook template directory key. Empty string means no compile hook. */
export const HOOK_TEMPLATE_MAP: Readonly<Record<string, string>> = {
  "java-maven": "java-maven",
  "java-gradle": "java-gradle",
  "kotlin-gradle": "kotlin",
  "typescript-npm": "typescript",
  "python-pip": "",
  "go-go": "go",
  "rust-cargo": "rust",
  "csharp-dotnet": "csharp",
};

/** Settings language JSON key (without .json). */
export const SETTINGS_LANG_MAP: Readonly<Record<string, string>> = {
  "java-maven": "java-maven",
  "java-gradle": "java-gradle",
  "kotlin-gradle": "java-gradle",
  "typescript-npm": "typescript-npm",
  "python-pip": "python-pip",
  "go-go": "go",
  "rust-cargo": "rust-cargo",
  "csharp-dotnet": "csharp-dotnet",
};

/** Database name to settings key. */
export const DATABASE_SETTINGS_MAP: Readonly<Record<string, string>> = {
  "postgresql": "database-psql",
  "mysql": "database-mysql",
  "oracle": "database-oracle",
  "mongodb": "database-mongodb",
  "cassandra": "database-cassandra",
};

/** Cache name to settings key. */
export const CACHE_SETTINGS_MAP: Readonly<Record<string, string>> = {
  "redis": "cache-redis",
  "dragonfly": "cache-dragonfly",
  "memcached": "cache-memcached",
};

/** Return hook template directory key, or empty string if none. */
export function getHookTemplateKey(language: string, buildTool: string): string {
  return HOOK_TEMPLATE_MAP[`${language}-${buildTool}`] ?? "";
}

/** Return settings language JSON key, or empty string if none. */
export function getSettingsLangKey(language: string, buildTool: string): string {
  return SETTINGS_LANG_MAP[`${language}-${buildTool}`] ?? "";
}

/** Return database settings JSON key, or empty string if none. */
export function getDatabaseSettingsKey(dbName: string): string {
  return DATABASE_SETTINGS_MAP[dbName] ?? "";
}

/** Return cache settings JSON key, or empty string if none. */
export function getCacheSettingsKey(cacheName: string): string {
  return CACHE_SETTINGS_MAP[cacheName] ?? "";
}
